Subject: [PATCH] Kirkwood: Add support for 6282 based QNAP devices

Add support for the QNAP NAS devices based on Marvell's 6282 Kirkwood
chip (TS-119P+, TS-219P+ and TS-419P+).  The differences to the 6281
based devices are:
 - Ethernet PHY address
 - GPIOs used for buttons (TS-119P+/TS-219P+)

Signed-off-by: Martin Michlmayr <tbm@cyrius.com>

--- a/arch/arm/mach-kirkwood/ts219-setup.c	2010-12-06 22:07:50.000000000 +0000
+++ b/arch/arm/mach-kirkwood/ts219-setup.c	2010-12-06 22:09:57.000000000 +0000
@@ -153,6 +153,8 @@
 	MPP15_GPIO,		/* USB Copy button */
 	MPP16_GPIO,		/* Reset button */
 	MPP36_GPIO,		/* RAM: 0: 256 MB, 1: 512 MB */
+	MPP37_GPIO,		/* Reset button (on devices with 88F6282) */
+	MPP43_GPIO,		/* USB Copy button (on devices with 88F6282) */
 	MPP44_GPIO,		/* Board ID: 0: TS-11x, 1: TS-21x */
 	0
 };
@@ -186,6 +188,8 @@
 
 static void __init qnap_ts219_init(void)
 {
+	u32 dev, rev;
+
 	/*
 	 * Basic setup. Needs to be called early.
 	 */
@@ -199,7 +203,15 @@
 	kirkwood_spi_init();
 	kirkwood_i2c_init();
 	i2c_register_board_info(0, &qnap_ts219_i2c_rtc, 1);
+
+	kirkwood_pcie_id(&dev, &rev);
+	if (dev == MV88F6282_DEV_ID) {
+		qnap_ts219_buttons[0].gpio = 43; /* USB Copy button */
+		qnap_ts219_buttons[1].gpio = 37; /* Reset button */
+		qnap_ts219_ge00_data.phy_addr = MV643XX_ETH_PHY_ADDR(0);
+	}
 	kirkwood_ge00_init(&qnap_ts219_ge00_data);
+
 	kirkwood_sata_init(&qnap_ts219_sata_data);
 	kirkwood_ehci_init();
 	platform_device_register(&qnap_ts219_button_device);
--- a/arch/arm/mach-kirkwood/ts41x-setup.c	2010-12-06 22:08:05.000000000 +0000
+++ b/arch/arm/mach-kirkwood/ts41x-setup.c	2010-12-06 22:11:54.000000000 +0000
@@ -215,6 +215,8 @@
 
 static void __init qnap_ts41x_init(void)
 {
+	u32 dev, rev;
+
 	/*
 	 * Basic setup. Needs to be called early.
 	 */
@@ -228,8 +230,15 @@
 	kirkwood_spi_init();
 	kirkwood_i2c_init();
 	i2c_register_board_info(0, &qnap_ts41x_i2c_rtc, 1);
+
+	kirkwood_pcie_id(&dev, &rev);
+	if (dev == MV88F6282_DEV_ID) {
+		qnap_ts41x_ge00_data.phy_addr = MV643XX_ETH_PHY_ADDR(0);
+		qnap_ts41x_ge01_data.phy_addr = MV643XX_ETH_PHY_ADDR(1);
+	}
 	kirkwood_ge00_init(&qnap_ts41x_ge00_data);
 	kirkwood_ge01_init(&qnap_ts41x_ge01_data);
+
 	kirkwood_sata_init(&qnap_ts41x_sata_data);
 	kirkwood_ehci_init();
 	platform_device_register(&qnap_ts41x_button_device);
